### PR TITLE
AI Control Keybinds

### PR DESCRIPTION
**Pull request by ampersand38**
**When merged this pull request will:**
- Add keybinds for controlling AI near players
- Watch Cursor: Makes all selected units watch the cursor location. Pressing this command with the cursor over a selected unit will cancel the watch target.
- Move To Cursor: Makes all selected units move to the cursor location. Units will return to formation and waypoint afterward.
- Toggle AI PATH: Makes all selected units start or stop moving. Does not affect aiming or shooting.
- https://youtu.be/TcPfTkfiB9o
- https://www.youtube.com/watch?v=ml0xGoXfGY4
- https://youtu.be/-KCMRES5ii8

### CODE REVIEW

**Review by neilzar — APPROVED** [review]
I nominate this for the best addition to Zeus award

File: addons/common/functions/fnc_getCursorPosition.sqf [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#include "script_component.hpp"
```
Comment (mharis001) [inline addons/common/functions/fnc_getCursorPosition.sqf:1]: This function is very similar to `getPosFromScreen`. We should just have one that satisfies all of the requirements.

File: addons/editor/initKeybinds.sqf [hunk 2]
```diff
@@ -82,3 +82,60 @@
...
+        curatorMouseOver params ["_type", "_entity", "_index"];
+        private _isCancelling = (_type == "OBJECT") && {_entity in SELECTED_OBJECTS};
+        private _gunners = (SELECTED_OBJECTS apply {gunner vehicle _x}) - [objNull];
+
+        private _cursorPosASL = call EFUNC(common,getCursorPosition);
+        {
+            [QEGVAR(common,doWatch), [_x, [ASLToAGL _cursorPosASL, objNull] select _isCancelling], _x] call CBA_fnc_targetEvent;
+        } forEach _gunners;
+
+        true // handled, prevents vanilla eject
```
Comment (mharis001) [inline addons/editor/initKeybinds.sqf:97]: Just `handled` since unbound by default. Here and below.

File: addons/editor/initKeybinds.sqf [hunk 3]
```diff
@@ -82,3 +82,60 @@
...
+        _drivers = _drivers arrayIntersect _drivers;
+
+        {
+            private _isAIEnalbedPATH = _x getVariable [QEGVAR(common,isAIEnalbedPATH), true];
+            private _AIEvent = [QEGVAR(common,enableAI), QEGVAR(common,disableAI)] select _isAIEnalbedPATH;
+            [_AIEvent, [_x, "PATH"], _x] call CBA_fnc_targetEvent;
+            _x setVariable [QEGVAR(common,isAIEnalbedPATH), !_isAIEnalbedPATH, true];
+            if (_isAIEnalbedPATH) then {
+                _disabled = _disabled + 1;
+            } else {;
```
Comment (mharis001) [inline addons/editor/initKeybinds.sqf:129]:
```suggestion
            } else {
```

File: addons/editor/initKeybinds.sqf [hunk 4]
```diff
@@ -82,3 +82,60 @@
...
+                _enabled = _enabled + 1;
+            };
+        } forEach _drivers;
+
+        private _message = "PATH " +
+            (["", format [" - disabled: %1", _disabled]] select (_disabled > 0)) +
+            (["", format [" - enabled: %1", _enabled]] select (_enabled > 0));
```
Comment (mharis001) [inline addons/editor/initKeybinds.sqf:136]: Localize.

File: addons/editor/initKeybinds.sqf [hunk 5]
```diff
@@ -82,3 +82,60 @@
...
+
+[ELSTRING(main,DisplayName), QGVAR(toggleAIPATH), [LSTRING(ToggleAIPATH), LSTRING(ToggleAIPATH_Description)], {
+    if (!isNull curatorCamera && {!GETMVAR(RscDisplayCurator_search,false) && {count SELECTED_OBJECTS > 0}}) then {
+        private _disabled = 0;
+        private _enabled = 0;
+        private _drivers = (SELECTED_OBJECTS apply {driver vehicle _x}) - [objNull];
+        _drivers = _drivers arrayIntersect _drivers;
+
+        {
+            private _isAIEnalbedPATH = _x getVariable [QEGVAR(common,isAIEnalbedPATH), true];
```
Comment (mharis001) [inline addons/editor/initKeybinds.sqf:123]: Incorrect spelling.

File: addons/editor/initKeybinds.sqf [hunk 6]
```diff
@@ -82,3 +82,60 @@
         };
     };
 }, {}, [0, [false, false, false]]] call CBA_fnc_addKeybind; // Default: Unbound
+
+[ELSTRING(main,DisplayName), QGVAR(watchCursor), [LSTRING(WatchCursor), LSTRING(WatchCursor_Description)], {
+    if (!isNull curatorCamera && {!GETMVAR(RscDisplayCurator_search,false)}) then {
+        curatorMouseOver params ["_type", "_entity", "_index"];
```
Comment (mharis001) [inline addons/editor/initKeybinds.sqf:88]: `_index` is not used.

File: addons/editor/initKeybinds.sqf [hunk 7]
```diff
@@ -82,3 +82,60 @@
...
+        private _isCancelling = (_type == "OBJECT") && {_entity in SELECTED_OBJECTS};
```
Comment (mharis001) [inline addons/editor/initKeybinds.sqf:89]: Unnecessary parentheses.

**Review by mharis001 — CHANGES_REQUESTED** [review]

File: addons/common/functions/fnc_getCursorPosition.sqf (hunk 1 above)
Comment (ampersand38) [reply addons/common/functions/fnc_getCursorPosition.sqf]: https://github.com/zen-mod/ZEN/pull/496#discussion_r558844225

File: addons/editor/initKeybinds.sqf (hunk 7 above)
Comment (ampersand38) [reply addons/editor/initKeybinds.sqf]: 943868e

File: addons/editor/initKeybinds.sqf (hunk 6 above)
Comment (ampersand38) [reply addons/editor/initKeybinds.sqf]: 839de51

File: addons/editor/initKeybinds.sqf (hunk 5 above)
Comment (ampersand38) [reply addons/editor/initKeybinds.sqf]: a79cd99

File: addons/editor/initKeybinds.sqf (hunk 4 above)
Comment (ampersand38) [reply addons/editor/initKeybinds.sqf]: 831a906

File: addons/editor/initKeybinds.sqf (hunk 2 above)
Comment (ampersand38) [reply addons/editor/initKeybinds.sqf]: f2046a9

File: addons/editor/initKeybinds.sqf [hunk 8]
```diff
@@ -82,3 +82,84 @@
...
+            [QEGVAR(common,doWatch), [_x, _pos], _x] call CBA_fnc_targetEvent;
+        } forEach _gunners;
+
+        true // handled
+    };
+}, {}, [0, [false, false, false]]] call CBA_fnc_addKeybind; // Default: Unbound
+
+[ELSTRING(main,DisplayName), QGVAR(watchCursor), [LSTRING(WatchCursor), LSTRING(WatchCursor_Description)], {
+    if (!isNull curatorCamera && {!GETMVAR(RscDisplayCurator_search,false) && {count SELECTED_OBJECTS > 0}}) then {
+        curatorMouseOver params ["_type", "_entity", ""];
```
Comment (neilzar) [inline addons/editor/initKeybinds.sqf:100]:
```suggestion
        curatorMouseOver params ["_type", "_entity"];
```
Can be removed.

File: addons/editor/stringtable.xml [hunk 9]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+ForceFireForceFire<?xml version="1.0" encoding="utf-8"?>
```
Comment (neilzar) [inline addons/editor/stringtable.xml:1]: That is not supposed to be there, I think.
```suggestion
<?xml version="1.0" encoding="utf-8"?>
```

**Review by neilzar — CHANGES_REQUESTED** [review]

File: addons/editor/initKeybinds.sqf [hunk 10]
```diff
@@ -82,3 +82,84 @@
...
+[ELSTRING(main,DisplayName), QGVAR(forceFire), [LSTRING(ForceFire), LSTRING(ForceFire_Description)], {
+    if (!isNull curatorCamera && {!GETMVAR(RscDisplayCurator_search,false) && {count SELECTED_OBJECTS > 0}}) then {
+        private _shooters = SELECTED_OBJECTS select {!isNull group _x};
+        [QEGVAR(common,ForceFire), [clientOwner, _shooters], _shooters] call CBA_fnc_targetEvent;
+
+        true // handled
+    };
+}, {
+    [QEGVAR(common,ForceFire), [clientOwner, []]] call CBA_fnc_localEvent;
+    [QEGVAR(common,ForceFire), [clientOwner, []]] call CBA_fnc_remoteEvent;
```
Comment (neilzar) [inline addons/editor/initKeybinds.sqf:122]:
```suggestion
    [QEGVAR(common,ForceFire), [clientOwner, []]] call CBA_fnc_globalEvent;
```
Should have the same effect, shouldn't it?

**Review by neilzar — CHANGES_REQUESTED** [review]

File: addons/editor/initKeybinds.sqf (hunk 10 above)
Comment (ampersand38) [reply addons/editor/initKeybinds.sqf:122]: Yes! idk why I couldn't find that one.

**Comment by CreepPork** [comment]
I'd love to see the diamond selector (e.g. `zen_common_fnc_selectPosition`) for drawing the selections where the unit should go.

File: addons/common/XEH_postInit.sqf [hunk 11]
```diff
@@ -153,6 +153,77 @@
     _unit enableIRLasers _mode;
 }] call CBA_fnc_addEventHandler;
 
+[QGVAR(forceFire), {
```
Comment (mharis001) [inline addons/common/XEH_postInit.sqf:156]: Move to a function.

File: addons/common/XEH_postInit.sqf [hunk 12]
```diff
@@ -153,6 +153,77 @@
...
+    if (_shooters isEqualTo []) then {
+        GVAR(forceFireCurators) = GVAR(forceFireCurators) - [_curatorClientID];
+    } else {
+        // track which curators are forcing fire
+        GVAR(forceFireCurators) pushBackUnique _curatorClientID;
+        [{
+            params ["_args", "_pfhID"];
+            _args params ["_curatorClientID", "_shooters", "_startTime"];
+
+            if (!(_curatorClientID in GVAR(forceFireCurators)) || {CBA_missionTime - _startTime > 10}) exitWith {
```
Comment (mharis001) [inline addons/common/XEH_postInit.sqf:169]: Change to `_endTime` so then the check is just `CBA_missionTime >= _endTime`.

**Review by mharis001 — COMMENTED** [review]
Previous this used `getCursorPosition` which was very similar to `getPosFromScreen` except that `getCursorPosition` did not check surface normal of intersections. If `getCursorPosition` produced better results for this then look to integrate its behaviour into `getPosFromScreen` using a parameter.

Also, there are a lot of keybinds now. We should consider moving these to a different component (maybe `ai` but `editor` is also fine) and/or categorizing these under a sub-category (maybe "AI Control").

**Comment by ampersand38** [comment]
> I'd love to see the diamond selector (e.g. `zen_common_fnc_selectPosition`) for drawing the selections where the unit should go.

Something like this? https://youtu.be/YnV08J_agek

File: addons/common/XEH_postInit.sqf (hunk 12 above)
Comment (ampersand38) [reply addons/common/XEH_postInit.sqf]: 32badb6

File: addons/editor/initKeybinds.sqf [hunk 13]
```diff
@@ -110,3 +110,82 @@
...
+
+[ELSTRING(main,DisplayName), QGVAR(toggleAIPATH), [LSTRING(ToggleAIPATH), LSTRING(ToggleAIPATH_Description)], {
+    if (!isNull curatorCamera && {!GETMVAR(RscDisplayCurator_search,false) && {count SELECTED_OBJECTS > 0}}) then {
+        private _disabled = 0;
+        private _enabled = 0;
+        private _drivers = (SELECTED_OBJECTS apply {driver vehicle _x}) - [objNull];
+        _drivers = _drivers arrayIntersect _drivers;
+
+        {
+            private _isAIEnabledPATH = _x getVariable [QEGVAR(common,isAIEnabledPATH), true];
```
[the remark this answers is not in the document]
Comment (ampersand38) [reply addons/editor/initKeybinds.sqf]: 8b3c25b

File: addons/editor/initKeybinds.sqf [hunk 14]
```diff
@@ -110,3 +110,82 @@
...
+            if (_isAIEnabledPATH) then {
+                _disabled = _disabled + 1;
+            } else {
+                _enabled = _enabled + 1;
+            };
+        } forEach _drivers;
+
+        private _message = "PATH " +
+            (["", format [" - %1: %2", localize ELSTRING(common,Disabled), _disabled]] select (_disabled > 0)) +
+            (["", format [" - %1: %2", localize ELSTRING(common,Enabled), _enabled]] select (_enabled > 0));
```
[the remark this answers is not in the document]
Comment (ampersand38) [reply addons/editor/initKeybinds.sqf]: fcdf540

File: addons/editor/initKeybinds.sqf [hunk 15]
```diff
@@ -110,3 +110,82 @@
...
+                [QEGVAR(common,doMove), [_x, ASLToAGL _cursorPosASL], _x] call CBA_fnc_targetEvent;
+            };
+        } forEach SELECTED_OBJECTS;
+
+        true // handled
+    };
+}, {}, [0, [false, false, false]]] call CBA_fnc_addKeybind; // Default: Unbound
+
+[ELSTRING(main,DisplayName), QGVAR(toggleAIPATH), [LSTRING(ToggleAIPATH), LSTRING(ToggleAIPATH_Description)], {
+    if (!isNull curatorCamera && {!GETMVAR(RscDisplayCurator_search,false) && {count SELECTED_OBJECTS > 0}}) then {
```
[the remark this answers is not in the document]
Comment (ampersand38) [reply addons/editor/initKeybinds.sqf]: 41b8a41

File: addons/common/XEH_postInit.sqf (hunk 11 above)
Comment (ampersand38) [reply addons/common/XEH_postInit.sqf]: a938a9f

File: addons/common/functions/fnc_drawHintLineIcon.sqf [hunk 16]
```diff
@@ -0,0 +1,61 @@
+#include "script_component.hpp"
```
Comment (mharis001) [inline addons/common/functions/fnc_drawHintLineIcon.sqf:1]: This function seems a bit too specific for these keybinds. I like the idea of common functions that allow for easy 3D hints. I think this should be split into 2 functions, one for line drawing and the other for icon drawing (`drawHintLine`, `drawHintIcon`).

Also, add support for objects as positions so lines and icons update as objects move.

File: addons/common/functions/fnc_drawHintLineIcon.sqf [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    if (_icon1 isNotEqualTo []) then {
+        _icon1 params [
+            ["_icon", ""],
+            ["_text", ""],
+            ["_scale", 1],
+            ["_angle", 0]
+        ];
+        drawIcon3D [_icon, _color, _position1, _scale, _scale, _angle, _text];
+    };
+}, 0, [CBA_missionTime + _timeShown, _position0, _position1, _icon0, _icon1, _color]] call CBA_fnc_addPerFrameHandler
```
Comment (mharis001) [inline addons/common/functions/fnc_drawHintLineIcon.sqf:61]: Let's use `Draw3D` EH instead.

File: addons/common/functions/fnc_forceFire.sqf [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+_shooters = _shooters select {
+    local _x
+    && {alive _x}
+    && {!isNull _x}
+    // Is vehicle, on foot, or in a turret
+    && {_x == vehicle _x || {_x call cba_fnc_turretPath isNotEqualTo []}}
+};
+
+if (_shooters isEqualTo []) exitWith {
+ GVAR(forceFireCurators) = GVAR(forceFireCurators) - [_curatorClientID];
```
Comment (mharis001) [inline addons/common/functions/fnc_forceFire.sqf:33]: Identation.

File: addons/common/functions/fnc_forceFire.sqf [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+        ) then {
+            if (_vehicle == _shooter) then {
+                // On foot
+                weaponState _shooter params ["", "_muzzle", "_firemode"];
+                _shooter forceWeaponFire [_muzzle, _firemode];
+            } else {
+                (fullCrew _vehicle select {_x select 0 == _shooter} select 0) params ["", "", "_cargoIndex", "_turretPath", "_isFFV"];
+                // FFV
+                if (_isFFV) exitWith {
+                    systemChat "FFV";
```
Comment (mharis001) [inline addons/common/functions/fnc_forceFire.sqf:69]: Remove.

File: addons/common/functions/fnc_forceFire.sqf (hunk 18 above)
Comment (ampersand38) [reply addons/common/functions/fnc_forceFire.sqf]: fe0b0a4

File: addons/common/functions/fnc_forceFire.sqf (hunk 19 above)
Comment (ampersand38) [reply addons/common/functions/fnc_forceFire.sqf]: c08db31

File: addons/common/functions/fnc_drawHintLineIcon.sqf (hunk 16 above)
Comment (ampersand38) [reply addons/common/functions/fnc_drawHintLineIcon.sqf]: How does this look? 8a91f43

File: addons/common/functions/fnc_forceFire.sqf [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+                // On foot
+                weaponState _shooter params ["", "_muzzle", "_firemode"];
+                _shooter forceWeaponFire [_muzzle, _firemode];
+            } else {
+                (fullCrew _vehicle select {_x select 0 == _shooter} select 0) params ["", "", "_cargoIndex", "_turretPath", "_isFFV"];
+                // FFV
+                if (_isFFV) exitWith {
+                    weaponState _shooter params ["", "_muzzle", "_firemode"];
+                    _shooter forceWeaponFire [_muzzle, _firemode];
+                };
```
Comment (mharis001) [inline addons/common/functions/fnc_forceFire.sqf:71]: Does not work for FFV units.

**Comment by IdoX1** [comment]
This is awesome!!
Would love to see it merged soon!

**Comment by ampersand38** [comment]
> Previously this used `getCursorPosition` which was very similar to `getPosFromScreen` except that `getCursorPosition` did not check surface normal of intersections. If `getCursorPosition` produced better results for this then look to integrate its behaviour into `getPosFromScreen` using a parameter.
> 
> Also, there are a lot of keybinds now. We should consider moving these to a different component (maybe `ai` but `editor` is also fine) and/or categorizing these under a sub-category (maybe "AI Control").

Intersections should be fine. Sometimes for an upper floor position the AI will just go below it, so it's better to make the AI move to a point above it, like on a wall.

Controls split into categories, I also grabbed the existing countermeasures key.
image

**Comment by ampersand38** [comment]
Perhaps I should take the hints into a separate PR?

**Comment by mharis001** [comment]
I pushed a `drawHint` common function that does 2D (map display) and 3D drawing and uses IDs to overwrite hints (so multiple hints are not shown for the same unit if the keybinds are used quickly - only the most recent one).

**Comment by ampersand38** [comment]
Looks great! I think it works fine with successive inputs. There's immediate feedback for the first one(s), which is then replaced by the next.

**Comment by mharis001** [comment]
The last thing I want to see done for this is remove the duplicated code between force fire and suppressive fire. Ideally, the common code would be extracted out to a separate function that both of these features can use.

**Review by neilzar — APPROVED** [review]

File: addons/common/functions/fnc_isUnitFFV.sqf [hunk 21]
```diff
@@ -18,3 +18,4 @@
 params [["_unit", objNull, [objNull]]];
 
 fullCrew vehicle _unit select {_x select 0 == _unit} param [0, []] param [4, false]
+&& {currentMuzzle _unit == weaponstate _unit select 1} // Tank commander is only FFV when turned out
```
Comment (mharis001) [inline addons/common/functions/fnc_isUnitFFV.sqf:21]: Would `CBA_fnc_isTurnedOut` work better here?

File: addons/common/functions/fnc_canFire.sqf [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+ * Public: No
+ */
+
+params [["_unit", objNull], ["_ignoreAmmo", false]];
+
+if (isNull _unit) exitWith {false};
+
+// If a vehicle is given directly, use the first turret that can fire
+if !(_unit isKindOf "CAManBase") exitWith {
+    [[_unit] call FUNC(firstTurretUnit)] call FUNC(canFire);
```
Comment (mharis001) [inline addons/common/functions/fnc_canFire.sqf:25]: `firstTurretUnit` also checks `canFire`. Will try to look at reworking this.

File: addons/common/functions/fnc_firstTurretUnit.sqf [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+
+if (isNull _vehicle) exitWith {objNull};
+
+private _firstTurretUnit = objNull;
+{
+    private _unit = _vehicle turretUnit _x;
+    if ([_unit] call FUNC(canFire)) exitWith {
+        _firstTurretUnit = _unit
+    };
+} forEach allTurrets _vehicle + [[-1]];
```
Comment (mharis001) [inline addons/common/functions/fnc_firstTurretUnit.sqf:28]: `getAllTurrets` common function.

**Review by mharis001 — CHANGES_REQUESTED** [review]

**Comment by mharis001** [comment]
I think this is finally good to go. I made the following changes and did some cleanup:

- Added an is reloading check to the can fire function so that it can effectively be used to time when to fire the next shot.
- Now using `CBA_fnc_isTurnedOut` for the turned out check in `isUnitFFV`.
- Moved the force fire function back to common and simplified it a bit (we can look at reworking this later). 
- Finally, the keybind now uses a global event on trigger as well as release just to avoid any potential issues with different locality between a vehicle and its turret.

@ampersand38 Please look over my changes and merge when ready.

**Review by mharis001 — APPROVED** [review]

**Pull request merged**